### PR TITLE
Extract file utility from table module.

### PR DESCRIPTION
**Pull request by cheme**
This PR extract file related methods from table module.
It also change the initial file capacity to align file size.

### CODE REVIEW

File: src/file.rs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+			path: filepath,
+			file: RwLock::new(file),
+			capacity: AtomicU64::new(capacity),
+			dirty: AtomicBool::new(false),
+			origin,
+		})
+	}
+
+	fn create_file(&self) -> Result<std::fs::File> {
+		log::debug!(target: "parity-db", "Created value table {}", self.origin);
```
Comment (cheme) [inline src/file.rs:111]: `origin` is only use for this log.

File: src/file.rs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+		let file = std::fs::OpenOptions::new().create(true).read(true).write(true).open(self.path.as_path())?;
+		disable_read_ahead(&file)?;
```
Comment (cheme) [inline src/file.rs:113]: In open file we preallocate, in create file we don't, maybe just remove preallocation or also preallocate in create file.

File: src/file.rs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+#[cfg(not(target_os = "macos"))]
+fn fsync(file: &std::fs::File) -> Result<()> {
+	file.sync_data()?;
+	Ok(())
+}
+
+
+const GROW_SIZE: u64 = 256 * 1024;
```
Comment (arkpar) [inline src/file.rs:74]: Should be `GROW_BYTES` or `GROW_SIZE_BYTES`

File: src/file.rs [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+pub struct DbFile {
```
Comment (arkpar) [inline src/file.rs:76]: Let's call it `TableFile`. `DbFile` sounds like a generic database file. This is specific for storing fixed size records though.

File: src/file.rs [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+	pub file: RwLock<Option<std::fs::File>>,
+	pub path: std::path::PathBuf,
+	pub capacity: AtomicU64,
+	pub dirty: AtomicBool,
+	pub origin: TableId,
+}
+
+impl DbFile {
+	pub fn open(filepath: std::path::PathBuf, header_size: u8, entry_size: u16, origin: TableId) -> Result<Self> {
```
Comment (arkpar) [inline src/file.rs:85]: Better name for `header_size` would be `header_entries`, as it is specified in the number of entries. Do we even need it to be a param, or is it always going to be 1?

File: src/file.rs [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+
+const GROW_SIZE: u64 = 256 * 1024;
+
+pub struct DbFile {
+	pub file: RwLock<Option<std::fs::File>>,
+	pub path: std::path::PathBuf,
+	pub capacity: AtomicU64,
+	pub dirty: AtomicBool,
+	pub origin: TableId,
```
Comment (arkpar) [inline src/file.rs:81]: Let's just call it `id`

File: src/file.rs (hunk 2 above)
Comment (arkpar) [reply src/file.rs:113]: open only preallocates to handle the case when opening an inconsistent database. I.e. there was a power failure and we ended up with a 0-sized file. To replay the log, we need to open the file as empty.
With normal operation `create_file` should be called from within the `grow` function only. Might be a good idea to remove the check for empty file from `write_at`

File: src/file.rs [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+impl DbFile {
+	pub fn open(filepath: std::path::PathBuf, header_size: u8, entry_size: u16, origin: TableId) -> Result<Self> {
+		let mut capacity = header_size as u64;
+		let file = if std::fs::metadata(&filepath).is_ok() {
+			let file = std::fs::OpenOptions::new().create(true).read(true).write(true).open(filepath.as_path())?;
+			disable_read_ahead(&file)?;
+			if file.metadata()?.len() == 0 {
+				// Preallocate.
+				debug_assert!(header_size as u64 * entry_size as u64 <= GROW_SIZE);
+				file.set_len(GROW_SIZE)?;
```
Comment (arkpar) [inline src/file.rs:93]: The file size should not be `GROW_SIZE` exactly, but rather `capacity` * `entry_size`. 
`GROW_SIZE` may contain a non-integer number of records.

**Review by arkpar — CHANGES_REQUESTED** [review]

File: src/file.rs (hunk 5 above)
Comment (cheme) [reply src/file.rs]: It's used my branch, but actually it is only for the debug assertion, I am removing it.

**Pull request merged**